### ISSUE TITLE
extendr_api::throw_r_error could leak a bit of memory

### ISSUE THREAD

**Issue by sorhawell** [eitsupi/prqlr#46]
Hello wonderful person, what a nice project! I personally never liked the raw SQL syntax variants ^^, maybe this is good bridge for me.

I noticed the use of [extendr_api::throw_r_error](https://github.com/extendr/extendr/issues/278#issuecomment-927620896). I guess in the case `prqlr` all the inputs or intermediate variabels of [the function compile()](https://github.com/eitsupi/prqlr/blob/1c5ba870ada95ca088b0f88f37c6ec10c0bdc9d8/src/rust/src/lib.rs#L12) could potentially be leaked. I suppose it is not large amounts of bytes as query strings are small. The linked thread have several alternative approaches, but no unified best approach for any use case.

```rust
pub fn compile(
    prql_query: &str,
    dialect: Option<String>,
    format: bool,
    signature_comment: bool,
) -> String {
    let dialect = prql_compiler::sql::Dialect::from_str(dialect.as_deref().unwrap_or_default())
        .map(From::from)
        .ok();

    let options: Option<prql_compiler::sql::Options> = Some(prql_compiler::sql::Options {
        format,
        dialect,
        signature_comment,
    });

    let result = Ok(prql_query)
        .and_then(prql_compiler::prql_to_pl)
        .and_then(prql_compiler::pl_to_rq)
        .and_then(|rq| prql_compiler::rq_to_sql(rq, options.map(prql_compiler::sql::Options::from)))
        .map_err(|e| e.composed("", prql_query, false));

    unwrap_or_throw(result)
}

fn unwrap_or_throw(result: anyhow::Result<String, prql_compiler::ErrorMessages>) -> String {
    match result {
        Ok(v) => v,
        Err(e) => {
            throw_r_error(e.to_string());
            unreachable!()
        }
    }
}
```

**Comment by eitsupi** [comment]
Hi, thanks for pointing that out!
I know very little Rust (and extendr and R....), so error handling is completely uncharted territory for me.......
I will take a look at that anyway.

**Comment by sorhawell** [comment]
With latest prqlr from r-univers it seems to leak mem 1:1-ratio with mem size of input string.
280 million character input string is quite beyond any realistic usecase though.

The 2nd value Vcells reflext reflect what amount of memory my task manager / resource monitor says is used by the rsession. gc() cannot free the memory because the rust function exited midway before dropping(running the rust destructor) the objects.


```r

> gc()
         used (Mb) gc trigger (Mb) limit (Mb) max used (Mb)
Ncells 507761 27.2    1116078 59.7         NA   669334 35.8
Vcells 814110  6.3    8388608 64.0      16384  1839428 14.1
> library(prqlr)
> rep(letters,1E7) |>
+     prql_compile() |>
+     cat()
Error in compile(prql_query, dialect, format, signature_comment) : 
  Expected Scalar, got Strings
> gc()
            used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
Ncells    516567   27.6    1116078   59.7         NA    694638   37.1
Vcells 260926302 1990.8  378169679 2885.3      16384 261023501 1991.5
> library(prqlr)
> rep(letters,1E7) |>
+     prql_compile() |>
+     cat()
Error in compile(prql_query, dialect, format, signature_comment) : 
  Expected Scalar, got Strings
> gc()
            used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
Ncells    519424   27.8    1116078   59.7         NA    985006   52.7
Vcells 520928899 3974.4  752725776 5742.9      16384 521114737 3975.8

Restarting R session...

> object.size(rep(letters,1E7))/1E6
2080 bytes "megabytes"

Restarting R session...

> gc()
         used (Mb) gc trigger (Mb) limit (Mb) max used (Mb)
Ncells 508080 27.2    1116990 59.7         NA   669334 35.8
Vcells 814864  6.3    8388608 64.0      16384  1839428 14.1
```

**Comment by sorhawell** [comment]
I will try to write a suggestion

**Comment by sorhawell** [comment]
My previous memory leakage test was likely inconclusive. I have used this test instead.

```r
library(prqlr)
for (i in 1:30) {
  x = paste(rep(letters,1E5),collapse ="")
  gc()
  print(tryCatch(prqlr::prql_compile(x), error = function(e) substr(as.character(e),1,10)))
}
```
I use "Activity monitor"(like windows task manager or similar) on a MacBook to assess leaked memory. Printed memory stats within an R session may under report leaked memory. I run the code snippet once, notice reserved memory for 'rsession' process thereafter. Run the code snippet again. Notice the reserved memory again. The leaked memory is the difference in reserved memory for the 'rsession' process between runs.

Current Mb leaked in above test given by branch:
main 1.1Gb
[drop_before_throw_error:](https://github.com/sorhawell/prqlr/tree/drop_before_throw_r_error) 100-250 Mb
[r_result_list:](https://github.com/sorhawell/prqlr/tree/r_result_list) 0 MB

drop_before_throw_error was an attempt to explicitly drop all objects before calling `throw_r_error( )`. I did not manage to completely avoid leakage.
r_result_list is copied from what I personally like to do. That is to pass a `Result<>` like object to the R side, and there have an `unwrap()`-Rfunction to do error handling. This just one of many proposed patterns discussed in issue [extendr issue 278](https://github.com/extendr/extendr/issues/278#issuecomment-927620896)

**Comment by eitsupi** [comment]
Thank you for working on this!
I read that issue and I don't like the way R lists are returned.
If you don't mind, could you create a PR with result list pattern?
I would like to merge.

**Comment by eitsupi** [comment]
> I'm confused which pattern. So not like r_result_list pattern right?

Oh, sorry, it's a typo! (corrected now)

#52 seems good.

**Comment by eitsupi** [comment]
Closed by #52
